pkg/blobinfocache/boltdb/boltdb.go

```diff
@@ -325,7 +325,67 @@ func (bdc *cache) appendReplacementCandidates(candidates []prioritize.CandidateW
 // data from previous RecordDigestUncompressedPair calls is used to also look up variants of the blob which have the same
 // uncompressed digest.
 func (bdc *cache) CandidateLocations2(transport types.ImageTransport, scope types.BICTransportScope, primaryDigest digest.Digest, canSubstitute bool) []blobinfocache.BICReplacementCandidate2 {
-	return bdc.candidateLocations(transport, scope, primaryDigest, canSubstitute, true)
+	resAllBlobs := []prioritize.CandidateWithTime{}
+	var uncompressedDigestValue digest.Digest // = ""
+	requireCompressionInfo := true
+	resultPrioritized := bdc.candidateLocations(transport, scope, primaryDigest, canSubstitute, requireCompressionInfo)
+	// Reprocess all blobs and append them to resultPrioritized in lower priority
+	if err := bdc.view(func(tx *bolt.Tx) error {
+		scopeBucket := tx.Bucket(knownLocationsBucket)
+		if scopeBucket == nil {
+			return nil
+		}
+		scopeBucket = scopeBucket.Bucket([]byte(transport.Name()))
+		if scopeBucket == nil {
+			return nil
+		}
+		if err := scopeBucket.ForEach(func(name []byte, v []byte) error {
```

Comment:
We don’t actually care about the `opaque` value, so it feels very wasteful to iterate over it.

If we we have a uncompressed -> digest association, this new code should return a _single_ new entry, not an entry for every registry that has that digest.

---

(Even without that, I’d be a bit unhappy about copy&pasting most of this loop nest. There’s enough logic in here that having a single parametrized implementation would be easier to maintain.)

Reply:
I am just thinking how can `appendReplacementCandidates` return the right candidates if i don't supply the top level bucket which i get by iterating over `opaque`, I must be missing something so I'll just try with the code itself and comment back here.

Reply:
`appendReplacementCandidates` uses `scopeBucket` for:
- Location, which we don’t need
- LastSeen, which I’m currently proposing we just don’t do. Alternatively, we could restructure this part of the database so that we only iterate over data for the relevant digest, not all the registries.

Reply:
Thanks, I was able to eliminate the scope loop by adding a new field `UnknownLocation` to `BICReplacementCandidate2` and modifying `appendReplacementCandidates`.

One doubt I think it still hard to replace/remove the below code from  `CandidateLocations2` since I was unable to find a way to get the digest corresponding to the uncompressedDigest without touching the bucket, but it ends in single iteration so i hope its not a blocker.
```go
b = b.Bucket([]byte(uncompressedDigestValue.String()))
   if b != nil {
        if err := b.ForEach(func(k, _ []byte) error {
	     d, err := digest.Parse(string(k))
```

Reply:
Sure, the uncompressed digest → compressed digest mapping does need to exist.